src/proposals/voice_management.cpp

```diff
@@ -46,11 +46,95 @@ ACTION dao::updatevoice (const uint64_t & start) {
       permission_level(get_self(), "active"_n),
       get_self(),
       "updatevoice"_n,
-      std::make_tuple(vitr->account.value)
+      std::make_tuple(vitr->account.value, scope)
     );
   }
 }
 
+
+ACTION dao::deletescope (const uint64_t & start, const name & scope) {
+
+  require_auth(get_self());
+
+  bool existing_scope = false;
+
+  for (auto & s : scopes) {
+    if (s == scope) {
+      existing_scope = true;
+    }
+  }
+  
+  check(existing_scope, "scope must exist");
+  
+  uint64_t batch_size = config_get(name("batchsize"));
+  uint64_t count = 0;
+
+  voice_tables voice_t(get_self(), scope.value);
+  auto vitr = start == 0 ? voice_t.begin() : voice_t.find(start);
+  
+  while (vitr != voice_t.end() && count < batch_size) {
+    vitr = voice_t.erase(vitr);
+    count++;
+  }
+  
+  if (vitr != voice_t.end()) {
+    send_deferred_transaction(
+      permission_level(get_self(), "active"_n),
+      get_self(),
+      "deletescope"_n,
+      std::make_tuple(vitr->account.value, scope)
+    );
+  }
+}
+
+ACTION dao::addvoice(const uint64_t & start, const name & scope) {
+
+  require_auth(get_self());
+
+  bool existing_scope = false;
+
+  for (auto & s : scopes) {
+    if (s == scope) { existing_scope = true; }
+  }
+
+  check(existing_scope, "scope must exist");
+
+  cs_points_tables cspoints_t(contracts::harvest, contracts::harvest.value);
+
+  voice_tables voices_t(get_self(), campaign_scope.value);
+  auto vitr = start == 0 ? voices_t.begin() : voices_t.find(start);
+
+  uint64_t batch_size = config_get(name("batchsize"));
+  uint64_t count = 0;
+
+  while (vitr != voices_t.end() && count < batch_size) {
+    auto csitr = cspoints_t.find(vitr->account.value);
+    uint64_t voice_amount = 0;
+
+    if (csitr != cspoints_t.end()) {
+      voice_amount = calculate_decay(csitr->rank);
+    }
+
+    print("account: ", vitr->account, ", points: ", voice_amount, scope);
+    set_voice(vitr->account, voice_amount, scope);
+
+    vitr++;
+    count++;
+  }
+
```

Comment:
Now that we added a new voice scope - how can the contract know which scopes exist?

We don't store the scopes anywhere, do we?

lets say I add a new voice scope called "foobar"

Then that voice scope won't get decayed, right?

Reply:
Well, there is a vector of scopes, also addvoices check for scopes in that vector